util/session/sessionmanager.go

```diff
@@ -152,11 +253,155 @@ func (mgr *SessionManager) Parse(tokenString string) (jwt.Claims, error) {
 	return token.Claims, nil
 }
 
+// GetLoginFailures retrieves the login failure information from the cache
+func (mgr *SessionManager) GetLoginFailures() map[string]LoginAttempts {
+	// Get failures from the cache
+	var failures map[string]LoginAttempts
+	err := mgr.cache.GetItem(loginAttemptsCacheKey, &failures)
+	if err != nil {
+		if err != appstate.ErrCacheMiss {
+			log.Errorf("Could not retrieve %s from cache: %s", loginAttemptsCacheKey, err.Error())
+		}
+		failures = make(map[string]LoginAttempts)
+	}
+
+	return failures
+}
+
+func expireOldFailedAttempts(maxAge time.Duration, failures *map[string]LoginAttempts) int {
+	expiredCount := 0
+	for key, attempt := range *failures {
+		if time.Since(attempt.LastFailed) > maxAge*time.Second {
+			expiredCount += 1
+			delete(*failures, key)
+		}
+	}
+	return expiredCount
+}
+
+// Updates the failure count for a given username. If failed is true, increases the counter. Otherwise, sets counter back to 0.
+func (mgr *SessionManager) updateFailureCount(username string, failed bool) {
+
+	failures := mgr.GetLoginFailures()
+
+	// Expire old entries in the cache if we have a failure window defined.
+	if window := getLoginFailureWindow(); window > 0 {
+		count := expireOldFailedAttempts(window, &failures)
+		if count > 0 {
+			log.Infof("Expired %d entries from session cache due to max age reached", count)
+		}
+	}
+
+	// If we exceed a certain cache size, we need to remove random entries to
+	// prevent overbloating the cache with fake entries, as this could lead to
+	// memory exhaustion and ultimately in a DoS. We remove a single entry to
+	// replace it with the new one.
+	//
+	// Chances are that we remove the one that is under active attack, but this
+	// chance is low (1:cache_size)
+	if failed && len(failures) >= getMaximumCacheSize() {
+		log.Warnf("Session cache size exceeds %d entries, removing random entry", getMaximumCacheSize())
+		idx := rand.Intn(len(failures) - 1)
+		var rmUser string
+		i := 0
+		for key := range failures {
+			if i == idx {
+				rmUser = key
+				delete(failures, key)
+				break
+			}
+			i++
+		}
+		log.Infof("Deleted entry for user %s from cache", rmUser)
+	}
+
+	attempt, ok := failures[username]
+	if !ok {
+		attempt = LoginAttempts{FailCount: 0}
+	}
+
+	// On login failure, increase fail count and update last failed timestamp.
+	// On login success, remove the entry from the cache.
+	if failed {
+		attempt.FailCount += 1
+		attempt.LastFailed = time.Now()
+		failures[username] = attempt
+		log.Warnf("User %s failed login %d time(s)", username, attempt.FailCount)
+	} else {
+		if attempt.FailCount > 0 {
+			// Forget username for cache size enforcement, since entry in cache was deleted
+			delete(failures, username)
+		}
+	}
+
+	err := mgr.cache.SetItem(loginAttemptsCacheKey, failures, 0, false)
```

Comment:
Cache getters and setters are located in https://github.com/argoproj/argo-cd/blob/da04075120e0e2e402e340c4f527cd3caf504c62/server/cache/cache.go#L43 . I think it makes sense to add GetLoginAttempts/SetLoginAttempts to `server/cache/Cache` and use it in session manager